torch/autograd/gradcheck.py

```diff
@@ -47,7 +47,77 @@ def iter_tensors(x: Union[torch.Tensor, Iterable[torch.Tensor]], only_requiring_
                 yield result
 
 
-def get_numerical_jacobian(fn, input, target=None, eps=1e-3, grad_out=1.0):
+def iter_tensor(x_tensor):
+    # Enumerates over a tensor and provides a corresponding flat index that translates
+    # to a given rol/col in the jacobian matrix. iter_tensor also returns a
```

Comment:
I've never read the get_numeric_jacobian before so this is a bit magical to me. How does the "jacobian matrix" in the example correspond to the "tensor"? What is an example of an iterator that this function outputs?

Reply:
For strided, `d_idx` corresponds to the correct column in the jacobian matrix. And the order is the same as as if we flatten a contiguous tensor. An example is that for a tensor t with size (2, 2): it will yield `x, 0, (0, 0)`, `x, 1, (0, 1)`, `x, 2, (1, 0)`, `x, 3, (1, 1)` where x is the `t.data` of the original tensor. The input t has numel 4, so Jacobian should have 4 columns. So having a `d_idx` of 3 and `idx` of `(1, 1)` indicates that perturbing `t[(1, 1)]` allows us to updating the third (last) column of any jacobian corresponding to this particular input.

Reply:
Mkldnn is similar, but requires us to do a dense copy first. Sparse is a bit more complicated since we need to reverse engineer the strides from the size, and also consider that it may have both sparse/dense dims.